backend/routeplanner/planner/domain/a_star.py

```diff
@@ -1,9 +1,149 @@
+
 import osmnx as ox
 import networkx as nx
+
+from . import router
 from .map import get_graph
+from .router import distance
+
+
+class AStarNode:
 
+    def __init__(self, node, parent: (), f):
+        self.node= node
+        self.parent = parent
+        self.g = 0  # Distance to start node
+        self.h = 0  # Distance to goal node
+        self.lc = 0  # Left turns cost
+        self.f = f # Total cost
+
+        # # Compare nodes
+        # def __eq__(self, other):
+        #     return self.node['osmid'] == other.node['osmid']
+
+# nie kończy nigdy wyszukiwać, nie wiem gdzie błąd
 def find_path(node1, node2):
-    # TODO: Implement a star algorithm here, function should return ordered list of nodes in path
-    # The path should not include first node
-    # graph = get_graph()
-    return []
\ No newline at end of file
+
+
+    # skąd graf ? wszystkie sąsiednie nody muszę być w grafie, aby obliczać g w 79 linii
+    graph= get_graph("Warszawa", 'drive')
+
+    # graph = ox.add_edge_speeds(graph, fallback = 50)
+    # graph = ox.add_edge_travel_times(graph)
+
+    node3=graph.nodes[ox.get_nearest_node(graph,(node1[1],node1[0]), method='haversine')]
+
+
+
+    node4 = graph.nodes[ox.get_nearest_node(graph, (node2[1], node2[0]), method='haversine')]
+
+
+
+    #route = nx.shortest_path(graph, node3['osmid'], node4['osmid'], weight='travel_time')
+
+
+
+
+
+    open =[]
+    closed =[]
+
+    start= AStarNode(node3,  None, 0)
+    end= AStarNode(node4,  None, 0)
+    open.append( start)
+
+
+    while len(open)>0:
+       # print("node")
+
+        open.sort(key=lambda x: x.f, reverse=False)
+
+        current_node =open.pop(0)
+        closed.append(current_node)
+        print(current_node.node)
+
+
+        if(current_node.node['osmid']== end.node['osmid']):
+            path = []
+            while current_node.node != start.node:
+                    path.append((current_node.node['x'], current_node.node['y']))
+                    print(current_node.node['osmid'])
+                    current_node = current_node.parent
+            return path[::-1]
+
+        neighbors= find_neighbors(current_node, graph)
+        for next in neighbors:
+
+            neighbor= AStarNode(graph.nodes[next], current_node.node, 0)
+            #neighbor.node musi być w grafie
+            neighbor.g = current_node.g + graph.edges[(current_node.node['osmid'], neighbor.node['osmid'], 0)]['travel_time']
+
+            neighbor.h =(distance((current_node.node['y'], current_node.node['x']), (end.node['y'], end.node['x'])))/60
```

Comment:
Dobra znalazłem spory błąd logiczny. Funkcja distance to odległość(nie czas) i zapisujesz ją do h a w g masz właśnie czas i na koniec to sumujesz 🤦‍♀️